### PR TITLE
feat(loggly): support HTTP response code to SYSLOG severity mapping

### LINKED ISSUE #6112: Proposal: Develop Loggly logging plugin
## Initial phase:
reference : [here](https://www.loggly.com/)
A logging plugin that works and does its job : ) 
Supported Protocol: SYSLOG-UDP

### schema:
```lua
local schema = {
    type = "object",
    properties = {
        customer_token = {type = "string"},
        severity = {
            type = "string",
            default = "INFO",
            enum = {"DEBUG", "INFO", "NOTICE", "WARNING", "ERR", "CRIT", "ALERT", "EMEGR",
                    "debug", "info", "notice", "warning", "err", "crit", "alert", "emegr"},
            description = "base severity log level",
        },
        include_req_body = {type = "boolean", default = false},
        include_resp_body = {type = "boolean", default = false},
        include_resp_body_expr = {
            type = "array",
            minItems = 1,
            items = {
                type = "array"
            }
        },
        tags = {
            type = "array",
            minItems = 1,
            items = {
                type = "string"
            }
        },
        prefer_name = {type = "boolean", default = true}
    },
    required = {"customer_token"}
}

```

### Metadata schema
```lua
local defaults = {
    host = "logs-01.loggly.com",
    port = 514,
    protocol = "syslog",
    timeout = 5000
}

local metadata_schema = {
    type = "object",
    properties = {
        host = {
            type = "string",
            default = defaults.host
        },
        port = {
            type = "integer",
            default = defaults.port
        },
        protocol = {
            type = "string",
            default = defaults.protocol,
            -- more methods coming soon
            enum = {"syslog"}
        },
        timeout = {
            type = "integer",
            minimum = 1,
            default= defaults.timeout
        },
        log_format = {
            type = "object",
        }
    }
}

```

## Phase-2
- Supported Protocol: HTTP[S]
- Support Changing Log Severity level based on HTTP response code.

This issue is for the tracking purpose of the plugin development. If you have any useful insight, feel free to let us know : )

### PR DESCRIPTION
**Pull request by bisakhmondal**
### What this PR does / why we need it:

closed #6112 (phase 2 step 2)
### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

**Review by tzssangglass — DISMISSED** [review]

File: apisix/plugins/loggly.lua [hunk 1]
```diff
@@ -184,9 +202,18 @@ local function generate_log_message(conf, ctx)
             core.table.insert(taglist, "tag=\"" .. conf.tags[i] .. "\"")
         end
     end
+
+    local message_severity = severity[conf.severity:upper()]
```
[the remark this answers is not in the document]
Comment (bisakhmondal) [reply apisix/plugins/loggly.lua:212]: Done

File: apisix/plugins/loggly.lua [hunk 2]
```diff
@@ -145,6 +150,19 @@ function _M.check_schema(conf, schema_type)
     if not ok then
         return nil, err
     end
+
+    if conf.severity_map then
+        for k, v in pairs(conf.severity_map) do
+            local rcode = tonumber(k)
```
[the remark this answers is not in the document]
Comment (bisakhmondal) [reply apisix/plugins/loggly.lua:164]: Done

File: apisix/plugins/loggly.lua [hunk 3]
```diff
@@ -145,6 +157,18 @@ function _M.check_schema(conf, schema_type)
...
     end
+
+    if conf.severity_map then
+        local cache = {}
+        for k, v in pairs(conf.severity_map) do
+            local rcode = tonumber(k)
+            if not rcode or rcode < 100 or rcode >= 600 then
+                return nil, "expecting severity_map with http response code([100,599]) as keys"
+            end
+            cache[k] = severity[v:upper()]
```
Comment (spacewander) [inline apisix/plugins/loggly.lua:168]: We can convert k to a number?

File: apisix/plugins/loggly.lua [hunk 4]
```diff
@@ -85,6 +86,17 @@ local schema = {
             type = "boolean",
             default = true
         },
+        severity_map = {
+            type = "object",
+            description = "upstream response code vs syslog severity mapping",
+            patternProperties = {
+                [".*"] = {
```
Comment (spacewander) [inline apisix/plugins/loggly.lua:93]: We can use regex to verify the range of status code?

Comment (bisakhmondal) [reply apisix/plugins/loggly.lua:93]: Yea, I had tested. It doesn't work that way. 
say for eg,
`["^[1-5][0-9]{2}"] = {...}` the check schema match the enums for only the set of keys where the key is a status code (regex). For other scenarios (key = "ABC", "500a") there is no matching regex, so it simply doesn't match and validate with the pattern and let it pass as valid inputs.

File: apisix/plugins/loggly.lua (hunk 3 above)
Comment (bisakhmondal) [reply apisix/plugins/loggly.lua:168]: I think I avoid that due to the default behaviour of cjson encoding `Cannot serialise table: excessively sparse array`. 
we can override the behaviour by `cjson.encode_sparse_array(true)`. In this case, I thought I don't need it that's why I am sticking with strings. WDYT?

Comment (spacewander) [reply apisix/plugins/loggly.lua:168]: OK

File: apisix/plugins/loggly.lua (hunk 4 above)
Comment (spacewander) [reply apisix/plugins/loggly.lua:93]: Does `["^[1-5][0-9]{2}$"] = {...}` work for it?

Comment (bisakhmondal) [reply apisix/plugins/loggly.lua:93]: @spacewander It works as well as doesn't work.
Example 1: Want :heavy_check_mark:  Got: :heavy_check_mark: 
```json
...
           "severity_map": {
                "200": "INFO"
            }
...
```

Example 2: Want :x:  Got: :x: 
```json
...
           "severity_map": {
                "200": "abc"
            }
...
```

Example 3: Want :x:  Got: :heavy_check_mark: : 
```json
...
           "severity_map": {
                "abc": "def"
            }
...
```
The problem is, if the key fails to match regex, the pattern properties don't get applied. Also, the program doesn't throw an error for the key mismatch.

Comment (spacewander) [reply apisix/plugins/loggly.lua:93]: Got it. What about adding `"additionalProperties": false`? So the mismatch key will cause an error.

Comment (bisakhmondal) [reply apisix/plugins/loggly.lua]: It worked! Awesome : )
```
// inp "severity_map": {
//               "abc": "def"
//            }
{
    "error_msg": "failed to check the configuration of plugin loggly err: property \"severity_map\" validation failed: additional properties forbidden, found abc"
}
```

**Review by spacewander — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**